framework/src/controller/application.js

```diff
@@ -224,6 +224,10 @@ class Application {
 		);
 		assert(Transaction, 'Transaction implementation is required');
 
+		if (options.matcher) {
+			Transaction.matcher = options.matcher;
+		}
```

Comment:
I would suggest this block to be changed as: 

```
const defaultMatcher = scope => scope.transaction.type === transactionType;

Transaction.matcher = options.matcher || Transaction.matcher || defaultMatcher; 
```

This way it will be more flexible for users to specify the matcher, first just as type, second as static property of class itself and on third level as custom matcher in options.

Reply:
The `scope` only contains the block basic informations ({ blockVersion, blockHeight, blockTimestamp })

Reply:
It should contain the raw JSON of the transaction as well, so user can have flexible matcher based on transaction content as well.

Reply:
transaction itself shouldn't  be used for matcher.
Only thing that can change the blockchain protocol should be block

Reply:
We may introduce new version of transactions, or transaction processing itself can be different for any particular transactions (as in our case exceptions), so passing the transaction JSON to the scope of matcher is very useful. 

At the end matcher just need to resolve the transaction class, no matter through which data it should do it. It does not impact the processing of transaction. The more flexible user facing interface could, the more use cases can be achieved by community (which we don't know right now).

Reply:
I think we shouldnt introduce something that we know it won't work...
one progress of blockchain is only happen by a block, and protocol change can only happen in the block.
Therefore, we should not introduce this at this point

Reply:
The matcher can't just be used for protocol change, it can be used to handle exceptions and other more generalised cases as well. And its not true that passing transaction JSON itself will break any thing. 

So I don't clearly agree to 

> I think we shouldnt introduce something that we know it won't work...

There is no reason that it would not work.

Reply:
Exception handling might happen in the similar way, but it will not be in this matcher property.
This issue is for design pattern for protocol change, and for this case, passing transaction is unnecessary and will not work because you cannot change protocol in the middle of a block

Reply:
In short, One use case of `matcher` is to handle protocol changes. It's not restricted only to that use case. So keep it flexible let others to came up with use cases which we can't think of right now.